unixpwd/c/unixpwd.c

```diff
@@ -0,0 +1,284 @@
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include <errno.h>
+#include <malloc.h>
+#include <pwd.h>
+#include <shadow.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <limits.h>
+
+#ifdef DEVELOPMENT
+#define ETC_PASSWD   "passwd"
+#define TMP_PASSWD   "passwd.XXXXXX"
+#define ETC_SPASSWD  "shadow"
+#define TMP_SPASSWD  "shadow.XXXXXX"
+#else
+#define ETC_PASSWD   "/etc/passwd"
+#define TMP_PASSWD   "/etc/passwd.XXXXXX"
+#define ETC_SPASSWD  "/etc/shadow"
+#define TMP_SPASSWD  "/etc/shadow.XXXXXX"
+#endif
+
+#define BUFLEN 4096
+
+char           *
+unixpwd_getpwd(const char *user)
+{
+    struct passwd   pwd,
+                   *pw;
+    char            buf[BUFLEN];
+
+    errno = 0;
+    if (getpwnam_r(user, &pwd, buf, BUFLEN, &pw) == 0 && pw)
+        return strdup(pw->pw_passwd);
+    if (errno == 0)
+        errno = EINVAL;
+    return NULL;
+}
+
+char           *
+unixpwd_getspw(const char *user)
+{
+    struct spwd     spw,
+                   *sp;
+    char            buf[BUFLEN];
+
+    if (getspnam_r(user, &spw, buf, BUFLEN, &sp) == 0 && sp)
+        return strdup(sp->sp_pwdp);
+    if (errno == 0)
+        errno = EINVAL;
+    return NULL;
+}
+
+
+char           *
+unixpwd_get(const char *user)
+{
+    struct spwd     spw,
+                   *sp;
+    struct passwd   pwd,
+                   *pw;
+    char            buf[BUFLEN];
+
+    errno = 0;
+    if (getspnam_r(user, &spw, buf, BUFLEN, &sp) == 0 && sp)
+        return strdup(sp->sp_pwdp);
+    if (getpwnam_r(user, &pwd, buf, BUFLEN, &pw) == 0 && pw)
+        return strdup(pw->pw_passwd);
+    if (errno == 0)
+        errno = EINVAL;
+    return NULL;
+}
+
+int
+unixpwd_setpwd(const char *user, char *password)
+{
+
+    struct passwd   pwd,
+                   *pw;
+    char            buf[BUFLEN];
+    int             tmp;
+    FILE           *tmp_file;
+    char            tmp_name[PATH_MAX];
+    struct stat     statbuf;
+    int             rc;
+    int             updated = 0;
+
+    strncpy(tmp_name, TMP_PASSWD, sizeof tmp_name);
+    tmp = mkstemp(tmp_name);
+    if (tmp == -1)
+        return errno;
+    if (stat(ETC_PASSWD, &statbuf) != 0) {
+        rc = errno;
+        close(tmp);
+        return rc;
+    }
+    if (fchown(tmp, statbuf.st_uid, statbuf.st_gid) != 0) {
+        rc = errno;
+        close(tmp);
+        return rc;
+    }
+    if (fchmod(tmp, statbuf.st_mode) != 0) {
+        rc = errno;
+        close(tmp);
```

Comment:
minor: the cleanup code is the same for all three of these, could they be part of the same if `stat (...) != 0 || fchown(...) != 0 || fchmod(...) != 0)`?
Or another way I used to write C code is to have something that resembles a try/finally:
```
rc = 0;
tmp = ...
do {
 if (... something fails... ) {
    rc = <code>;
    break;
  }
 ...
} while(0);
close(tmp);
return rc;
```
(or errno instead of rc)

Reply:
I agree that there is some repetitiveness in the code that would be good to reduce. I like the first suggestion.